File: exercism/python/complex_numbers.py
import math

class ComplexNumber:
    def __init__(self, real=0, imaginary=0):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)
        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __radd__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)
        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __mul__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)
        return ComplexNumber(self.real * other.real - self.imaginary * other.imaginary, self.real * other.imaginary + self.imaginary * other.real)

    def __rmul__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)
        return ComplexNumber(self.real * other.real - self.imaginary * other.imaginary, self.real * other.imaginary + self.imaginary * other.real)

    def __sub__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)
        return ComplexNumber(self.real - other.real, self.imaginary - other.imaginary)

    def __rsub__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)
        return ComplexNumber(other.real - self.real, other.imaginary - self.imaginary)

    def __truediv__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)

        conjugate = other.conjugate()
        denominator = (other * conjugate).real
        numerator = self * conjugate

        return ComplexNumber(numerator.real / denominator, numerator.imaginary / denominator)

    def __rtruediv__(self, other):
        if not isinstance(other, ComplexNumber):
            other = ComplexNumber(other)

        conjugate = self.conjugate()
        denominator = self * conjugate
        numerator = other * conjugate

        return ComplexNumber(numerator.real / denominator.real, numerator.imaginary / denominator.real)

    def __abs__(self):
        return math.sqrt(self.real**2 + self.imaginary**2)

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        return math.exp(self.real) * ComplexNumber(math.cos(self.imaginary), math.sin(self.imaginary))

File: exercism/haskell/ComplexNumbers.hs
module ComplexNumbers
(Complex,
 conjugate,
 abs,
 exp,
 real,
 imaginary,
 mul,
 add,
 sub,
 div,
 complex) where

import Prelude hiding (div, abs, exp)
import qualified Prelude as P (exp)

-- NOTE: Implementing Num and Fractional instances would be better,
-- giving us other functions for free, like (^)

-- Data definition -------------------------------------------------------------
data Complex a = a :+ a deriving(Eq, Show)

complex :: (a, a) -> Complex a
complex (x, y) = x :+ y

-- unary operators -------------------------------------------------------------
conjugate :: Num a => Complex a -> Complex a
conjugate (x :+ y) = x :+ (-y)

abs :: Floating a => Complex a -> a
abs (x :+ y) = sqrt $ x**2 + y**2

real :: Num a => Complex a -> a
real (x :+ _) = x

imaginary :: Num a => Complex a -> a
imaginary (_ :+ y) = y

exp :: Floating a => Complex a -> Complex a
exp (x :+ y) =
  let n = P.exp x
   in (n * cos y) :+ (n * sin y)

-- binary operators ------------------------------------------------------------
mul :: Num a => Complex a -> Complex a -> Complex a
mul (x :+ y) (a :+ b) = (x * a - y * b) :+ (x * b + y * a)

add :: Num a => Complex a -> Complex a -> Complex a
add (x :+ y) (a :+ b) = (x + a) :+ (y + b)

sub :: Num a => Complex a -> Complex a -> Complex a
sub (x :+ y) (a :+ b) = (x - a) :+ (y - b)

div :: Fractional a => Complex a -> Complex a -> Complex a
div (x :+ y) (a :+ b) =
  let n = (a * a + b * b)
   in ((x * a + y * b)/n) :+ ((y * a - x * b)/n)
